### ISSUE TITLE
Error java.util.ConcurrentModificationException into JSONUtils

### ISSUE THREAD

**Issue by silassilvadev** [OneSignal/OneSignal-Android-SDK#1086]
The error was reported by reports in an Analytics. Unfortunately it was not possible to reproduce the error.

```java
java.util.ConcurrentModificationException: 
	at java.util.LinkedHashMap$LinkedHashIterator.nextEntry()(LinkedHashMap.java:346)
	at java.util.LinkedHashMap$KeyIterator.next()(LinkedHashMap.java:366)
	at com.onesignal.JSONUtils.generateJsonDiff()(JSONUtils.java:43)
	at com.onesignal.UserState.generateJsonDiff()(UserState.java:252)
	at com.onesignal.UserState.generateJsonDiff()(UserState.java:115)
	at com.onesignal.UserStateSynchronizer.persist()(UserStateSynchronizer.java:192)
	at com.onesignal.OneSignalStateSynchronizer.persist()(OneSignalStateSynchronizer.java:79)
	at com.onesignal.OneSignal.scheduleSyncService()(OneSignal.java:1253)
	at com.onesignal.OneSignal.onAppLostFocus()(OneSignal.java:1248)
	at com.onesignal.ActivityLifecycleHandler$AppFocusRunnable.run()(ActivityLifecycleHandler.java:263)
	at android.os.Handler.handleCallback()(Handler.java:739)
	at android.os.Handler.dispatchMessage()(Handler.java:95)
	at android.os.Looper.loop()(Looper.java:148)
	at android.os.HandlerThread.run()(HandlerThread.java:61)
```

OneSignal Version: 3.14.1
Versions: 5.0.1, 5.1, 5.1.1, 6, 6.0.1, 7, 7.1.1, 7.1.2, 8, 8.1, 9 e 10 .
Devices: Samsung, Motorola, LG, Redmi, Asus, Xioami e OnePlus.

**Comment by jkasten2** [comment]
@silassilvadev Thanks for reporting. How many users is the effecting?

We will look into this issue. However to  address the issue quicker I would recommend accessing OneSignal from the same thread if possible.

**Comment by silassilvadev** [comment]
@jkasten2 Thanks for the feedback. There are 1733 users affected in the last 3 days. The direct calls I make to OneSignal come in three stages: library initialization, user registration, external id registration and device id capture. None of these calls seem to be linked to the error found, and checking here I realized that the calls are being made in the same thread.

**Comment by Jeasmine** [comment]
Hey, @silassilvadev can you try the last 3.15.2 version? There are several bug fixes

**Comment by silassilvadev** [comment]
@Jeasmine I haven't updated yet, because I haven't seen a release note about the problem. Is there a fix for this problem in the new version?

**Comment by Jeasmine** [comment]
This issue has not been reproduced, we had some thread issues going on though, so if you could try the new version that will help to discard those same problems

**Comment by rgomezp** [comment]
This has been reported in as recent as 3.15.1

**Comment by jkasten2** [comment]
Stack traces from version 3.15.1
```Java
Fatal Exception: java.util.ConcurrentModificationException
  at java.util.LinkedHashMap$LinkedHashIterator.nextEntry(LinkedHashMap.java:411)
  at java.util.LinkedHashMap$KeyIterator.next(LinkedHashMap.java:422)
  at com.onesignal.JSONUtils.generateJsonDiff(JSONUtils.java:43)
  at com.onesignal.UserState.generateJsonDiff(UserState.java:260)
  at com.onesignal.UserState.generateJsonDiff(UserState.java:123)
  at com.onesignal.UserStateSynchronizer.internalSyncUserState(UserStateSynchronizer.java:231)
  at com.onesignal.UserStateSynchronizer.syncUserState(UserStateSynchronizer.java:213)
  at com.onesignal.UserStateSynchronizer$NetworkHandlerThread$1.run(UserStateSynchronizer.java:110)
  at android.os.Handler.handleCallback(Handler.java:761)
  at android.os.Handler.dispatchMessage(Handler.java:98)
  at android.os.Looper.loop(Looper.java:156)
  at android.os.HandlerThread.run(HandlerThread.java:61)
```

```Java
Fatal Exception: java.util.ConcurrentModificationException
  at java.util.LinkedHashMap$LinkedHashIterator.nextNode(LinkedHashMap.java:760)
  at java.util.LinkedHashMap$LinkedEntryIterator.next(LinkedHashMap.java:793)
  at java.util.LinkedHashMap$LinkedEntryIterator.next(LinkedHashMap.java:791)
  at org.json.JSONObject.writeTo(JSONObject.java:722)
  at org.json.JSONObject.toString(JSONObject.java:692)
  at com.onesignal.UserState.persistState(UserState.java:203)
  at com.onesignal.UserStateSynchronizer.internalSyncUserState(UserStateSynchronizer.java:241)
  at com.onesignal.UserStateSynchronizer.syncUserState(UserStateSynchronizer.java:213)
  at com.onesignal.UserStateSynchronizer$NetworkHandlerThread$1.run(UserStateSynchronizer.java:110)
  at android.os.Handler.handleCallback(Handler.java:907)
  at android.os.Handler.dispatchMessage(Handler.java:105)
  at android.os.Looper.loop(Looper.java:216)
  at android.os.HandlerThread.run(HandlerThread.java:65)
```

**Comment by Jeasmine** [comment]
Fixed on last release 3.15.5 under https://github.com/OneSignal/OneSignal-Android-SDK/pull/1203

**Comment by matteinn** [comment]
@Jeasmine do you know when will 3.15.5 published to maven?
https://mvnrepository.com/artifact/com.onesignal/OneSignal/3.15.5
